File: go-cplr/conformance_test.go
package sduicompiler

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"gopkg.in/yaml.v3"
)

const specDir = "../spec"

var fixturesDir = filepath.Join(specDir, "fixtures")

type versioningQuery struct {
	AppVersion *string `yaml:"app_version"`
	Template   string  `yaml:"template"`
}

type conformanceCases struct {
	Golden     []string `yaml:"golden"`
	Versioning struct {
		Fixture            string            `yaml:"fixture"`
		Screen             string            `yaml:"screen"`
		Queries            []versioningQuery `yaml:"queries"`
		InvalidAppVersions []string          `yaml:"invalid_app_versions"`
	} `yaml:"versioning"`
	Errors map[string]string `yaml:"errors"`
}

func loadCases(t *testing.T) conformanceCases {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(specDir, "cases.yaml"))
	if err != nil {
		t.Fatalf("read cases.yaml: %v", err)
	}
	var cases conformanceCases
	if err := yaml.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse cases.yaml: %v", err)
	}
	return cases
}

func resolveTemplateVersion(t *testing.T, registry *ScreenRegistry, id, templateVersion string) *ComposedTemplate {
	t.Helper()
	module := registry.ModuleOf(id)
	if module == nil {
		t.Fatalf("Unknown screen: %s", id)
	}
	threshold := ""
	for _, entry := range module.Versions {
		if entry.Assets.Template == templateVersion {
			threshold = entry.Version
			break
		}
	}
	if threshold == "" {
		t.Fatalf("No threshold maps to template %s", templateVersion)
	}
	composed, err := registry.Resolve(id, threshold)
	if err != nil {
		t.Fatalf("resolve %s@%s: %v", id, threshold, err)
	}
	if composed == nil {
		t.Fatalf("Failed to resolve %s", id)
	}
	return composed
}

func TestConformanceGolden(t *testing.T) {
	cases := loadCases(t)
	for _, fixture := range cases.Golden {
		t.Run(fixture, func(t *testing.T) {
			input := filepath.Join(fixturesDir, fixture, "input")
			expectedDir := filepath.Join(fixturesDir, fixture, "expected")
			registry, err := BuildScreenRegistry(input)
			if err != nil {
				t.Fatalf("build: %v", err)
			}

			manifestBytes, err := os.ReadFile(filepath.Join(expectedDir, "manifest.json"))
			if err != nil {
				t.Fatalf("read expected manifest: %v", err)
			}
			var manifest map[string]struct {
				Etags map[string]string `json:"etags"`
			}
			if err := json.Unmarshal(manifestBytes, &manifest); err != nil {
				t.Fatalf("parse expected manifest: %v", err)
			}

			for id, entry := range manifest {
				screenDir := filepath.Join(expectedDir, "screens", id)
				files, err := os.ReadDir(screenDir)
				if err != nil {
					t.Fatalf("read %s: %v", screenDir, err)
				}
				for _, file := range files {
					templateVersion := strings.TrimSuffix(file.Name(), ".json")
					composed := resolveTemplateVersion(t, registry, id, templateVersion)
					expected, err := os.ReadFile(filepath.Join(screenDir, file.Name()))
					if err != nil {
						t.Fatalf("read expected screen: %v", err)
					}
					if got := CompactJSON(composed.Template); got != string(expected) {
						t.Errorf("screen %s@%s JSON mismatch:\n got: %s\nwant: %s", id, templateVersion, got, expected)
					}
					if composed.Etag != entry.Etags[templateVersion] {
						t.Errorf("screen %s@%s etag = %s, want %s", id, templateVersion, composed.Etag, entry.Etags[templateVersion])
					}
				}
			}
		})
	}
}

func TestConformanceVersioning(t *testing.T) {
	cases := loadCases(t)
	spec := cases.Versioning
	input := filepath.Join(fixturesDir, spec.Fixture, "input")

	for _, query := range spec.Queries {
		label := "(none)"
		appVersion := ""
		if query.AppVersion != nil {
			label = *query.AppVersion
			appVersion = *query.AppVersion
		}
		t.Run("selects "+query.Template+" for "+label, func(t *testing.T) {
			registry, err := BuildScreenRegistry(input)
			if err != nil {
				t.Fatalf("build: %v", err)
			}
			composed, err := registry.Resolve(spec.Screen, appVersion)
			if err != nil {
				t.Fatalf("resolve: %v", err)
			}
			viaThreshold := resolveTemplateVersion(t, registry, spec.Screen, query.Template)
			if composed == nil || composed.Etag != viaThreshold.Etag {
				t.Errorf("resolve(%s, %q) selected wrong template", spec.Screen, appVersion)
			}
		})
	}

	for _, invalid := range spec.InvalidAppVersions {
		t.Run("rejects "+invalid, func(t *testing.T) {
			registry, err := BuildScreenRegistry(input)
			if err != nil {
				t.Fatalf("build: %v", err)
			}
			if _, err := registry.Resolve(spec.Screen, invalid); err == nil || !strings.Contains(err.Error(), "Invalid semantic version") {
				t.Errorf("resolve(%s, %q) error = %v, want Invalid semantic version", spec.Screen, invalid, err)
			}
		})
	}
}

func TestConformanceErrors(t *testing.T) {
	cases := loadCases(t)
	for fixture, message := range cases.Errors {
		t.Run(fixture, func(t *testing.T) {
			input := filepath.Join(fixturesDir, "errors", fixture, "input")
			_, err := BuildScreenRegistry(input)
			if err == nil || !strings.Contains(err.Error(), message) {
				t.Errorf("build error = %v, want message containing %q", err, message)
			}
		})
	}
}

File: js-cplr/test/conformance.test.ts
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { load as loadYaml } from 'js-yaml';
import { describe, expect, it } from 'vitest';
import { ScreenRegistry } from '../src/registry.js';

const SPEC_DIR = join(import.meta.dirname, '..', '..', 'spec');
const FIXTURES_DIR = join(SPEC_DIR, 'fixtures');

interface VersioningQuery {
  app_version: string | null;
  template: string;
}

interface Cases {
  golden: string[];
  versioning: {
    fixture: string;
    screen: string;
    queries: VersioningQuery[];
    invalid_app_versions: string[];
  };
  errors: Record<string, string>;
}

const cases = loadYaml(
  readFileSync(join(SPEC_DIR, 'cases.yaml'), 'utf8'),
) as Cases;

describe('conformance', () => {
  describe('golden', () => {
    for (const fixture of cases.golden) {
      it(`composes ${fixture} byte-identically`, () => {
        const input = join(FIXTURES_DIR, fixture, 'input');
        const expectedDir = join(FIXTURES_DIR, fixture, 'expected');
        const registry = ScreenRegistry.build(input);
        const manifest = JSON.parse(
          readFileSync(join(expectedDir, 'manifest.json'), 'utf8'),
        ) as Record<string, { etags: Record<string, string> }>;

        for (const id of Object.keys(manifest)) {
          const screenDir = join(expectedDir, 'screens', id);
          for (const file of readdirSync(screenDir)) {
            const templateVersion = file.replace(/\.json$/, '');
            const composed = resolveTemplateVersion(
              registry,
              id,
              templateVersion,
            );
            expect(JSON.stringify(composed.template)).toBe(
              readFileSync(join(screenDir, file), 'utf8'),
            );
            expect(composed.etag).toBe(manifest[id].etags[templateVersion]);
          }
        }
      });
    }
  });

  describe('versioning', () => {
    const spec = cases.versioning;
    const input = join(FIXTURES_DIR, spec.fixture, 'input');

    for (const query of spec.queries) {
      it(`selects ${query.template} for app version ${query.app_version ?? '(none)'}`, () => {
        const registry = ScreenRegistry.build(input);
        const composed = registry.resolve(
          spec.screen,
          query.app_version ?? undefined,
        );
        const viaThreshold = resolveTemplateVersion(
          registry,
          spec.screen,
          query.template,
        );
        expect(composed?.etag).toBe(viaThreshold.etag);
      });
    }

    for (const invalid of spec.invalid_app_versions) {
      it(`rejects malformed app version ${invalid}`, () => {
        const registry = ScreenRegistry.build(input);
        expect(() => registry.resolve(spec.screen, invalid)).toThrow(
          'Invalid semantic version',
        );
      });
    }
  });

  describe('errors', () => {
    for (const [fixture, message] of Object.entries(cases.errors)) {
      it(`rejects ${fixture} mentioning "${message}"`, () => {
        const input = join(FIXTURES_DIR, 'errors', fixture, 'input');
        expect(() => ScreenRegistry.build(input)).toThrow(message);
      });
    }
  });
});

function resolveTemplateVersion(
  registry: ScreenRegistry,
  id: string,
  templateVersion: string,
) {
  const module = registry.moduleOf(id);
  if (module === undefined) throw new Error(`Unknown screen: ${id}`);
  const threshold = Object.entries(module.versions).find(
    ([, assets]) => assets.template === templateVersion,
  );
  if (threshold === undefined) {
    throw new Error(`No threshold maps to template ${templateVersion}`);
  }
  const composed = registry.resolve(id, threshold[0]);
  if (composed === undefined) throw new Error(`Failed to resolve ${id}`);
  return composed;
}
